### PR TITLE
Use warp-wide decoupled look-back in `multibyte_split`

### PR DESCRIPTION
**Pull request by upsj**
## Description
~~To finish off the trifecta of tuning PRs:
Our custom decoupled look-back uses a single thread instead of a warp (like cub does). This increases the look-back latency significantly. Compared to #11681 we get another almost 2x speedup by matching cub:~~

**I am not able to reproduce these results any more, it seems like the GPU I was running on is a bit slow, the speedup came from accidentally running on another GPU.**

<details>

| source_type | delim_size | delim_percent |    size_approx    |  GPU Time  | Peak Memory Usage | Encoded file size |
|-------------|------------|---------------|-------------------|------------|-------------------|-------------------|
|           0 |          1 |             1 | 2^30 = 1073741824 |  63.348 ms |         3.947 GiB |      1006.638 MiB |
|           0 |          4 |             1 | 2^30 = 1073741824 | 107.686 ms |         3.709 GiB |      1014.442 MiB |
|           0 |          7 |             1 | 2^30 = 1073741824 | 108.465 ms |         3.694 GiB |      1015.606 MiB |
|           0 |          1 |            25 | 2^30 = 1073741824 | 110.083 ms |         9.978 GiB |       762.462 MiB |
|           0 |          4 |            25 | 2^30 = 1073741824 | 136.302 ms |         5.292 GiB |       953.075 MiB |
|           0 |          7 |            25 | 2^30 = 1073741824 | 134.580 ms |         5.170 GiB |       980.321 MiB |

</details>

This probably needs a bit of scrutiny in terms of memory ordering, since I'm not deeply familiar with the guarantees the compiler provides w.r.t. cub::ThreadStore and libcuda++ atomics.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/src/io/text/multibyte_split.cu [hunk 1]
```diff
@@ -217,9 +220,9 @@ __global__ void multibyte_split_seed_kernel(
   cutoff_offset tile_output_offset)
 {
   auto const thread_idx = blockIdx.x * blockDim.x + threadIdx.x;
-  if (thread_idx == 0) {
-    tile_multistates.set_inclusive_prefix(-1, tile_multistate_seed);
-    tile_output_offsets.set_inclusive_prefix(-1, tile_output_offset);
+  if (thread_idx < 32) {
+    tile_multistates.set_inclusive_prefix(thread_idx - 32, tile_multistate_seed);
+    tile_output_offsets.set_inclusive_prefix(thread_idx - 32, tile_output_offset);
```
Comment (harrism) [inline cpp/src/io/text/multibyte_split.cu:225]: 32 is a magic number. I do think CUDA has a `constexpr` somewhere for the warp size that you should use.

Second, by definition the first argument (tile index) to `set_inclusive_prefix` will always be negative, resulting in a negative offset added to a pointer if `num_tiles` is also less than 32, IIUC.  Is that what is intended? Is `num_tiles > 32` a precondition for this code? If so, I think that should be documented and / or asserted.

**Review by harrism — COMMENTED** [review]
I don't know `multibyte_split` well enough to review in detail. Not much stood out except for `-32`.

File: cpp/src/io/text/multibyte_split.cu (hunk 1 above)
Comment (upsj) [reply cpp/src/io/text/multibyte_split.cu:225]: I couldn't find one, but it seems like a sensible constant to have. The tiles are arranged as a circular array, so negative indices wrap around. `num_tiles > 32` is assumed, I'll add an assertion for it.

Comment (harrism) [reply cpp/src/io/text/multibyte_split.cu:225]: https://github.com/rapidsai/cudf/blob/bf2c75113b2e0f146a5f5e8df22c244d5a39bdf3/cpp/include/cudf/detail/utilities/cuda.cuh#L36

Comment (upsj) [reply cpp/src/io/text/multibyte_split.cu:225]: thanks! I grepped for `= 32` only, there are just too many ways to initialize variables 😄

**Comment by upsj** [comment]
The performance differences vanish when checking on a consistent device, so this PR is unnecessary

**Pull request closed without merging**